# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path = os.getcwd()
print(path)

/home/joe/Desktop/GuidedCapstone/Step4_and_5


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv(path+'/../Step3/data/step3_output.csv')
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [4]:
df.describe()

summit_elev  vertical_drop  trams  fastEight  fastSixes   fastQuads  \
count    167.000000     167.000000  167.0      167.0      167.0  167.000000   
mean    3603.856287     835.497006    0.0        0.0        0.0    0.221557   
std     3236.111943     583.574495    0.0        0.0        0.0    0.563992   
min      315.000000     175.000000    0.0        0.0        0.0    0.000000   
25%     1240.000000     357.500000    0.0        0.0        0.0    0.000000   
50%     2000.000000     650.000000    0.0        0.0        0.0    0.000000   
75%     5779.500000    1150.000000    0.0        0.0        0.0    0.000000   
max    12075.000000    2600.000000    0.0        0.0        0.0    2.000000   

             quad      triple      double     surface  ...  SkiableTerrain_ac  \
count  167.000000  167.000000  167.000000  167.000000  ...         167.000000   
mean     0.538922    1.125749    1.658683    2.221557  ...         243.106594   
std      0.709221    1.172885    1.374345    1.420170  ...         322.663730   
min      0.000000    0.000000    0.000000    0.000000  ...          15.000000   
25%      0.000000    0.000000    1.000000    1.000000  ...          50.000000   
50%      0.000000    1.000000    1.000000    2.000000  ...         115.000000   
75%      1.000000    2.000000    2.000000    3.000000  ...         245.000000   
max      2.000000    5.000000    6.000000    6.000000  ...        1704.000000   

       Snow Making_ac  daysOpenLastYear   yearsOpen  averageSnowfall  \
count      167.000000        167.000000  167.000000       167.000000   
mean       102.222063        105.653403   59.219500       144.975442   
std         68.306294         21.946731   14.330361       104.562439   
min         13.000000         56.000000   22.000000        18.000000   
25%         43.500000         93.000000   51.000000        53.500000   
50%         84.000000        110.000000   58.000000       120.000000   
75%        174.873239        115.103943   67.500000       207.000000   
max        300.000000        163.000000   95.000000       500.000000   

       AdultWeekday  AdultWeekend  projectedDaysOpen  NightSkiing_ac  \
count    167.000000    167.000000         167.000000      167.000000   
mean      49.848378     56.406439         111.702545       41.634731   
std       15.831880     15.305689          20.862934       43.794339   
min       20.000000     20.000000          56.000000        0.000000   
25%       39.000000     45.000000         100.000000        0.000000   
50%       49.000000     55.000000         115.000000       35.000000   
75%       57.916957     65.000000         120.053004       69.500000   
max       89.000000     93.000000         180.000000      180.000000   

         clusters  
count  167.000000  
mean     0.550898  
std      0.749768  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      2.000000  

[8 rows x 24 columns]

In [5]:
#removing useless columns with all 0 values
df.drop(['trams','fastEight','fastSixes'], axis=1, inplace=True)

In [6]:
df.describe()

summit_elev  vertical_drop   fastQuads        quad      triple  \
count    167.000000     167.000000  167.000000  167.000000  167.000000   
mean    3603.856287     835.497006    0.221557    0.538922    1.125749   
std     3236.111943     583.574495    0.563992    0.709221    1.172885   
min      315.000000     175.000000    0.000000    0.000000    0.000000   
25%     1240.000000     357.500000    0.000000    0.000000    0.000000   
50%     2000.000000     650.000000    0.000000    0.000000    1.000000   
75%     5779.500000    1150.000000    0.000000    1.000000    2.000000   
max    12075.000000    2600.000000    2.000000    2.000000    5.000000   

           double     surface  total_chairs        Runs  TerrainParks  ...  \
count  167.000000  167.000000    167.000000  167.000000    167.000000  ...   
mean     1.658683    2.221557      5.766467   29.335329      2.122422  ...   
std      1.374345    1.420170      2.485975   17.148751      0.942253  ...   
min      0.000000    0.000000      0.000000    7.000000      1.000000  ...   
25%      1.000000    1.000000      4.000000   15.500000      1.000000  ...   
50%      1.000000    2.000000      5.000000   24.000000      2.000000  ...   
75%      2.000000    3.000000      7.000000   41.500000      2.820789  ...   
max      6.000000    6.000000     14.000000   83.000000      4.000000  ...   

       SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear   yearsOpen  \
count         167.000000      167.000000        167.000000  167.000000   
mean          243.106594      102.222063        105.653403   59.219500   
std           322.663730       68.306294         21.946731   14.330361   
min            15.000000       13.000000         56.000000   22.000000   
25%            50.000000       43.500000         93.000000   51.000000   
50%           115.000000       84.000000        110.000000   58.000000   
75%           245.000000      174.873239        115.103943   67.500000   
max          1704.000000      300.000000        163.000000   95.000000   

       averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
count       167.000000    167.000000    167.000000         167.000000   
mean        144.975442     49.848378     56.406439         111.702545   
std         104.562439     15.831880     15.305689          20.862934   
min          18.000000     20.000000     20.000000          56.000000   
25%          53.500000     39.000000     45.000000         100.000000   
50%         120.000000     49.000000     55.000000         115.000000   
75%         207.000000     57.916957     65.000000         120.053004   
max         500.000000     89.000000     93.000000         180.000000   

       NightSkiing_ac    clusters  
count      167.000000  167.000000  
mean        41.634731    0.550898  
std         43.794339    0.749768  
min          0.000000    0.000000  
25%          0.000000    0.000000  
50%         35.000000    0.000000  
75%         69.500000    1.000000  
max        180.000000    2.000000  

[8 rows x 21 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [7]:
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df.state)], axis=1)

In [8]:
df.head()

Name  summit_elev  vertical_drop  fastQuads  quad  \
0               Hilltop Ski Area         2090            294          0     0   
1            Sunrise Park Resort        11100           1800          1     2   
2  Yosemite Ski & Snowboard Area         7800            600          0     0   
3               Donner Ski Ranch         8012            750          0     0   
4                  June Mountain        10090           2590          2     0   

   triple  double  surface  total_chairs  Runs  ...  Rhode Island  \
0       1       0        2             3  13.0  ...             0   
1       3       1        0             7  65.0  ...             0   
2       1       3        1             5  10.0  ...             0   
3       1       5        2             8  52.0  ...             0   
4       0       4        1             7  41.0  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 57 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler()
scaler.fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [14]:
print('explained variance: ' + str(explained_variance_score(y_test, y_pred)))
print('mean absolute error: ' + str(mean_absolute_error(y_test, y_pred)))

explained variance: -3.4871197191798737e+23
mean absolute error: 2142520422208.8645


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [15]:
print(lm.intercept_)
'''some of the coefficients in my regression (including this intercept) ended up unreasonably large.
All X-values were standardized to a z-value and y values were between 0 and 100 so I don't know why
coefficients as large as 10^11 showed up. I'm going to continue through the rest of the project
rather than spend too much time trying to debug/understand why this happened at this point in time.
'''

-538837471396.12244


"some of the coefficients in my regression (including this intercept) ended up unreasonably large.\nAll X-values were standardized to a z-value and y values were between 0 and 100 so I don't know why\ncoefficients as large as 10^11 showed up. I'm going to continue through the rest of the project\nrather than spend too much time trying to debug/understand why this happened at this point in time.\n"

In [16]:
print(y_pred)

[ 5.13072510e+01  5.11802979e+01  5.08736572e+01  6.42711182e+01
  5.72603760e+01  6.80260010e+01  5.51241455e+01  4.16861572e+01
  4.07540283e+01  6.82174072e+01  5.57003174e+01  4.53614502e+01
  3.52012939e+01  5.13209229e+01  4.65142822e+01  6.02135010e+01
  4.03892822e+01  7.05172119e+01  7.34742432e+01 -2.99955875e+13
  5.72857666e+01  4.44176025e+01 -2.99946788e+13 -2.99955915e+13
  6.31666260e+01  6.01759033e+01  5.67427979e+01  4.82515869e+01
  6.81099854e+01  6.79244385e+01  3.19459229e+01  5.73052979e+01
  6.55963135e+01  7.10894775e+01  6.17115479e+01  3.89752197e+01
  7.28189697e+01  3.29844971e+01  5.40020752e+01  8.03736572e+01
  5.61636963e+01  4.09244385e+01]


In [17]:
print(y_test)

[47.         49.         68.         59.         59.         71.
 58.         55.         54.         65.         62.         48.
 42.         42.         44.         65.         39.         69.
 79.         64.16681004 55.         45.         79.         64.16681004
 64.         45.         59.         47.         75.         60.
 39.         45.         68.         64.16681004 57.         40.
 72.         34.         35.         92.         59.         46.        ]


In [18]:
pd.DataFrame(y_pred).describe()
#y_pred values are large because they're computed from my regression's coefficients which are large

0
count  4.200000e+01
mean  -2.142520e+12
std    7.818607e+12
min   -2.999559e+13
25%    4.236902e+01
50%    5.541223e+01
75%    6.399500e+01
max    8.037366e+01

In [19]:
X_with_coefs = pd.DataFrame(X_scaled.copy())
X_with_coefs.loc['coefs']=model.coef_
X_with_coefs
#a little debugging to see what went wrong. Not all coefficients are large, but many are.

0         1             2             3             4   \
0     -0.469208 -0.930688 -3.940186e-01 -7.621647e-01 -1.075355e-01   
1      2.323371  1.657721  1.384390e+00  2.066313e+00  1.602791e+00   
2      1.300562 -0.404756 -3.940186e-01 -7.621647e-01 -1.075355e-01   
3      1.366270 -0.146946 -3.940186e-01 -7.621647e-01 -1.075355e-01   
4      2.010329  3.015519  3.162798e+00 -7.621647e-01 -9.626985e-01   
...         ...       ...           ...           ...           ...   
163    1.185263 -0.043822 -3.940186e-01 -7.621647e-01 -1.075355e-01   
164    1.303041  1.264132 -3.940186e-01  6.520742e-01 -1.075355e-01   
165    1.877984  0.265549 -3.940186e-01 -7.621647e-01 -1.075355e-01   
166    1.827464  0.454609 -3.940186e-01 -7.621647e-01  7.476275e-01   
coefs  2.256288  0.640799  4.252908e+12  5.348040e+12  8.844403e+12   

                 5             6             7         8         9   ...  \
0     -1.210519e+00 -1.564765e-01 -1.116176e+00 -0.955432 -1.194793  ...   
1     -4.807116e-01 -1.568994e+00  4.976891e-01  2.085979 -0.130315  ...   
2      9.789036e-01 -8.627355e-01 -3.092437e-01 -1.130898 -0.130315  ...   
3      2.438519e+00 -1.564765e-01  9.011555e-01  1.325626 -0.130315  ...   
4      1.708711e+00 -8.627355e-01  4.976891e-01  0.682251 -0.130315  ...   
...             ...           ...           ...       ...       ...  ...   
163   -4.807116e-01 -8.627355e-01 -1.116176e+00  1.091672 -1.194793  ...   
164   -4.807116e-01 -1.568994e+00 -1.116176e+00  0.155853 -0.130315  ...   
165    9.789036e-01 -8.627355e-01 -3.092437e-01  0.214342 -0.130315  ...   
166   -1.210519e+00 -1.568994e+00 -1.519643e+00 -0.253568  0.743396  ...   
coefs  1.036356e+13  1.070911e+13 -1.874606e+13  2.082109  0.358398  ...   

               45            46            47            48            49  \
0       -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
1       -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
2       -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
3       -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
4       -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
...           ...           ...           ...           ...           ...   
163     -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
164     -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
165     -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
166     -0.077615 -7.761505e-02 -7.761505e-02 -1.352504e-01 -2.091650e-01   
coefs -575.727037  2.314169e+12  2.314169e+12  3.984038e+12  6.011045e+12   

                 50            51            52            53            54  
0     -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01 -1.756821e-01  
1     -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01 -1.756821e-01  
2     -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01 -1.756821e-01  
3     -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01 -1.756821e-01  
4     -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01 -1.756821e-01  
...             ...           ...           ...           ...           ...  
163   -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01  5.692100e+00  
164   -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01  5.692100e+00  
165   -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01  5.692100e+00  
166   -1.100964e-01 -1.566521e-01 -1.352504e-01 -2.523772e-01  5.692100e+00  
coefs  3.262856e+12  4.586324e+12  3.984038e+12  7.116899e+12  5.111913e+12  

[168 rows x 55 columns]

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coefficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [20]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
total_chairs       1.874606e+13
surface            1.070911e+13
double             1.036356e+13
New York           9.524641e+12
triple             8.844403e+12
Michigan           8.576466e+12
New Hampshire      7.440457e+12
Pennsylvania       7.116899e+12
Wisconsin          7.116899e+12
Minnesota          6.405964e+12
California         6.405964e+12
Vermont            6.011045e+12
Idaho              5.582510e+12
New Mexico         5.582510e+12
quad               5.348040e+12
Connecticut        5.111913e+12
Maine              5.111913e+12
Colorado           5.111913e+12
Wyoming            5.111913e+12
Massachusetts      4.586324e+12
Washington         4.586324e+12
Ohio               4.586324e+12
fastQuads          4.252908e+12
North Carolina     3.984038e+12
West Virginia      3.984038e+12
Illinois           3.984038e+12
Iowa               3.984038e+12
Oregon             3.984038e+12
Utah               3.984038e+12
Montana            3.984038e+12
Virginia           3.262856e+12
Missouri           3.262856e+12
Indiana            3.262856e+12
Tennessee          2.314169e+12
Nevada             2.314169e+12
Alaska             2.314169e+12
Arizona            2.314169e+12
South Dakota       2.314169e+12
Maryland           7.041146e+07
New Jersey         3.088409e+05
Rhode Island       5.757270e+02
AdultWeekday       1.026270e+01
summit_elev        2.256288e+00
Runs               2.082109e+00
NightSkiing_ac     1.244141e+00
clusters           9.362793e-01
LongestRun_mi      7.573242e-01
vertical_drop      6.407987e-01
daysOpenLastYear   4.921875e-01
SkiableTerrain_ac  4.902344e-01
yearsOpen          4.863281e-01
projectedDaysOpen  3.896484e-01
TerrainParks       3.583984e-01
averageSnowfall    1.755371e-01
Snow Making_ac     5.346680e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [21]:
#dropping all columns with state info
df_no_state = df.iloc[:,0:22]

In [22]:
df_no_state.head()

Name  summit_elev  vertical_drop  fastQuads  quad  \
0               Hilltop Ski Area         2090            294          0     0   
1            Sunrise Park Resort        11100           1800          1     2   
2  Yosemite Ski & Snowboard Area         7800            600          0     0   
3               Donner Ski Ranch         8012            750          0     0   
4                  June Mountain        10090           2590          2     0   

   triple  double  surface  total_chairs  Runs  ...  SkiableTerrain_ac  \
0       1       0        2             3  13.0  ...               30.0   
1       3       1        0             7  65.0  ...              800.0   
2       1       3        1             5  10.0  ...               88.0   
3       1       5        2             8  52.0  ...              505.0   
4       0       4        1             7  41.0  ...             1500.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0       30.000000        150.000000       36.0             69.0     30.000000   
1       80.000000        115.000000       49.0            250.0     74.000000   
2      174.873239        110.000000       84.0            300.0     47.000000   
3       60.000000        163.000000       82.0            400.0     75.000000   
4      174.873239        115.103943       58.0            250.0     57.916957   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0      34.00000              152.0            30.0         0  
1      78.00000              104.0            80.0         1  
2      47.00000              107.0             0.0         1  
3      75.00000              170.0             0.0         1  
4      64.16681              128.0             0.0         1  

[5 rows x 22 columns]

In [23]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_no_state.drop(['Name','AdultWeekend'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler()
scaler.fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [24]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [25]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [26]:
print('explained variance: ' + str(explained_variance_score(y_test, y_pred)))
print('mean absolute error: ' + str(mean_absolute_error(y_test, y_pred)))
print('y-intercept: '+str(lm.intercept_))

explained variance: 0.5559565165902216
mean absolute error: 6.852990686240759
y-intercept: 56.241296374899306


In [27]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         10.986869
summit_elev           3.408969
clusters              2.933922
daysOpenLastYear      2.063798
averageSnowfall       1.916504
Runs                  1.563800
vertical_drop         1.417618
triple                1.327128
surface               1.229705
NightSkiing_ac        0.804283
double                0.675288
projectedDaysOpen     0.626373
total_chairs          0.429953
quad                  0.418139
LongestRun_mi         0.243637
Snow Making_ac        0.222380
yearsOpen             0.131799
SkiableTerrain_ac     0.121466
fastQuads             0.060362
TerrainParks          0.025213

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [28]:
#I don't have base_elev feature, so only dropping summit_elev
df_no_state_elev = df_no_state.drop(axis=1, labels=['summit_elev'])

In [29]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_no_state_elev.drop(['Name','AdultWeekend'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler()
scaler.fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [30]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [31]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [32]:
print('explained variance: ' + str(explained_variance_score(y_test, y_pred)))
print('mean absolute error: ' + str(mean_absolute_error(y_test, y_pred)))
print('y-intercept: '+str(lm.intercept_))

explained variance: 0.5413080297878097
mean absolute error: 7.275361831392762
y-intercept: 56.169716762833055


In [33]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         10.718374
averageSnowfall       2.856805
clusters              2.234782
vertical_drop         1.905925
SkiableTerrain_ac     1.761634
daysOpenLastYear      1.459799
triple                1.229914
Runs                  1.099987
NightSkiing_ac        1.028189
surface               0.901449
double                0.861119
total_chairs          0.620436
quad                  0.278725
Snow Making_ac        0.262785
TerrainParks          0.230052
projectedDaysOpen     0.140035
LongestRun_mi         0.055306
yearsOpen             0.027533
fastQuads             0.001950

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -3.4871197191798737e+23 | 2142520422208.8645 |-|
| Model 2. | 0.5559565165902216|6.852990686240759 |'state'|
| Model 3. | 0.5413080297878097 |7.275361831392762 |'state','summit_elev','base_elev'|

Model Selection:

I will choose Model 2. Model 2 looks to be the best predictor as it has the lowest mean absolute error.